src/ast/types.ts

```diff
@@ -0,0 +1,70 @@
+export interface BaseNode {
+  // Every leaf interface that extends BaseNode must specify
+  // a string literal as a type property. This allows us to
+  // make use of TypeScript's discriminated unions.
+  type: string;
+}
+
+export interface Program extends BaseNode {
+  type: 'Program';
+  body: ExternalDeclaration[];
+}
+
+export type ExternalDeclaration = FunctionDeclaration | VariableDeclaration;
+
+export interface BaseStatement extends BaseNode {}
+
+export type Expression = AssignmentExpression | Identifier | Literal;
+
+export interface BaseExpression extends BaseNode {}
+
+export interface Identifier extends BaseExpression {
+  type: 'Identifier';
+  name: string;
+}
+
+export interface Literal extends BaseExpression {
```

Comment:
I'm not sure about using `Literal` because we may not handle strings as a whole. Maybe we can follow the BNF constants.

Reply:
Agreed. I'm removing it for this PR since the exact shape it should take isn't obvious until we actually start converting the CST nodes to it.